Knative deployments do not generate proper volume mounts for secrets volumes

**Describe the bug**
When deploying as a knative service, the kubernetes/openshift extension does not generate secrets volume mounts properly.

**Expected behavior**
The declared secrets are mounted in the container running as a knative service

**Actual behavior**
```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:1.8.0.Final:build (default) on project infinispan-client-quickstart: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR]         [error]: Build step io.quarkus.kubernetes.deployment.KubernetesDeployer#deploy threw an exception: io.dekorate.deps.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://api.cluster-462c.462c.example.opentlc.com:6443/apis/serving.knative.dev/v1/namespaces/dgdemo/services. Message: admission webhook "validation.webhook.serving.knative.dev" denied the request: validation failed: volumeMount has no matching volume: spec.template.spec.containers[0].volumeMounts[0].name. Received status: Status(apiVersion=v1, code=400, details=null, kind=Status, message=admission webhook "validation.webhook.serving.knative.dev" denied the request: validation failed: volumeMount has no matching volume: spec.template.spec.containers[0].volumeMounts[0].name, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=BadRequest, status=Failure, additionalProperties={}).
```

**To Reproduce**
Steps to reproduce the behavior:
1. Add the following to `application.properties`:

```
quarkus.kubernetes.deployment-target=knative
quarkus.knative.mounts.my-volume.path=/mnt
quarkus.knative.secret-volumes.my-volume.secret-name=clientcerts
```
Along with other necessary bits to deploy a Quarkus app as a knative service

2. Run `mvn clean package -Pnative -DskipTests` 
3. Observe failure above
4. Observe values in `target/kubernetes/knative.yml` show that the volume mount is properly declared:

```
        volumeMounts:
        - mountPath: /mnt
          name: my-volume
          readOnly: false
          subPath: ""

```

**But there is no corresponding `volume:` declared on the `Service` object. **

When deploying as a regular non-knative deployment I do see the volume in `openshift.yml`:

```
      volumes:
      - name: my-volume
        secret:
          defaultMode: 384
          optional: false
          secretName: clientcerts

```


**Configuration**
```properties
quarkus.infinispan-client.server-list=example-infinispan:11222

# Auth
quarkus.infinispan-client.auth-server-name=example-infinispan
quarkus.infinispan-client.auth-realm=default
quarkus.infinispan-client.auth-username=developer
quarkus.infinispan-client.auth-password=XXXXXXX
quarkus.infinispan-client.sasl-mechanism=PLAIN
quarkus.infinispan-client.client-intelligence=BASIC

# cert stuff
quarkus.infinispan-client.trust-store=/mnt/clientcerts
quarkus.infinispan-client.trust-store-password=password
quarkus.infinispan-client.trust-store-type=jks

quarkus.openshift.mounts.my-volume.path=/mnt
quarkus.openshift.secret-volumes.my-volume.secret-name=clientcerts

quarkus.knative.mounts.my-volume.path=/mnt
quarkus.knative.secret-volumes.my-volume.secret-name=clientcerts


quarkus.native.container-build=true
quarkus.native.builder-image=quay.io/quarkus/ubi-quarkus-mandrel:20.1-java11
quarkus.container-image.build=true
quarkus.kubernetes.deploy=true
quarkus.openshift.expose=true
quarkus.kubernetes-client.trust-certs=true
quarkus.kubernetes.deployment-target=knative
quarkus.container-image.registry=image-registry.openshift-image-registry.svc:5000
quarkus.container-image.group=dgdemo

```

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: `Darwin jfalkner-OSX 19.6.0 Darwin Kernel Version 19.6.0: Thu Jun 18 20:49:00 PDT 2020; root:xnu-6153.141.1~1/RELEASE_X86_64 x86_64`
 - Output of `java -version`: 

```
openjdk version "11.0.6" 2020-01-14
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.6+10)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.6+10, mixed mode)
```

 - GraalVM version (if different from Java): 

using Mandrel

 - Quarkus version or git rev: `1.8.0.Final`

 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): 
```
Apache Maven 3.6.2 (40f52333136460af0dc0d7232c0dc0bcf0d9e117; 2019-08-27T11:06:16-04:00)
Maven home: /Users/jfalkner/.m2/wrapper/dists/apache-maven-3.6.2-bin/795eh28tki48bv3l67maojf0ra/apache-maven-3.6.2
Java version: 11.0.6, vendor: AdoptOpenJDK, runtime: /Library/Java/JavaVirtualMachines/adoptopenjdk-11.jdk/Contents/Home
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "10.15.6", arch: "x86_64", family: "mac"
```

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -152,7 +152,7 @@
         <aws-alexa-sdk.version>2.37.1</aws-alexa-sdk.version>
         <azure-functions-java-library.version>1.3.0</azure-functions-java-library.version>
         <kotlin.version>1.4.20</kotlin.version>
-        <dekorate.version>0.14.0</dekorate.version>
+        <dekorate.version>0.14.2</dekorate.version>
         <maven-artifact-transfer.version>0.10.0</maven-artifact-transfer.version>
         <jline.version>2.14.6</jline.version>
         <maven-invoker.version>3.0.1</maven-invoker.version>

--- a/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KnativeProcessor.java
+++ b/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KnativeProcessor.java
@@ -9,6 +9,14 @@ import java.util.List;
 import java.util.Optional;
 import java.util.stream.Stream;
 
+import io.dekorate.knative.decorator.AddAwsElasticBlockStoreVolumeToRevisionDecorator;
+import io.dekorate.knative.decorator.AddAzureDiskVolumeToRevisionDecorator;
+import io.dekorate.knative.decorator.AddAzureFileVolumeToRevisionDecorator;
+import io.dekorate.knative.decorator.AddConfigMapVolumeToRevisionDecorator;
+import io.dekorate.knative.decorator.AddHostAliasesToRevisionDecorator;
+import io.dekorate.knative.decorator.AddPvcVolumeToRevisionDecorator;
+import io.dekorate.knative.decorator.AddSecretVolumeToRevisionDecorator;
+import io.dekorate.knative.decorator.AddSidecarToRevisionDecorator;
 import io.dekorate.knative.decorator.ApplyGlobalAutoscalingClassDecorator;
 import io.dekorate.knative.decorator.ApplyGlobalContainerConcurrencyDecorator;
 import io.dekorate.knative.decorator.ApplyGlobalRequestsPerSecondTargetDecorator;
@@ -210,6 +218,52 @@ public class KnativeProcessor {
             long percent = traffic.percent.orElse(100L);
             result.add(new DecoratorBuildItem(KNATIVE,
                     new ApplyTrafficDecorator(name, revisionName, latestRevision, percent, tag)));
+        });
+
+        //Add revision decorators
+        result.addAll(createVolumeDecorators(project, name, config));
+        config.getHostAliases().entrySet().forEach(e -> {
+            result.add(new DecoratorBuildItem(KNATIVE,
+                    new AddHostAliasesToRevisionDecorator(name, HostAliasConverter.convert(e))));
+        });
+        config.getSidecars().entrySet().forEach(e -> {
+            result.add(new DecoratorBuildItem(KNATIVE, new AddSidecarToRevisionDecorator(name, ContainerConverter.convert(e))));
+        });
+
+        return result;
+    }
+
+    private static List<DecoratorBuildItem> createVolumeDecorators(Optional<Project> project, String name,
+            PlatformConfiguration config) {
+        List<DecoratorBuildItem> result = new ArrayList<>();
+
+        config.getSecretVolumes().entrySet().forEach(e -> {
+            result.add(
+                    new DecoratorBuildItem(KNATIVE, new AddSecretVolumeToRevisionDecorator(SecretVolumeConverter.convert(e))));
+        });
+
+        config.getConfigMapVolumes().entrySet().forEach(e -> {
+            result.add(new DecoratorBuildItem(KNATIVE,
+                    new AddConfigMapVolumeToRevisionDecorator(ConfigMapVolumeConverter.convert(e))));
+        });
+
+        config.getPvcVolumes().entrySet().forEach(e -> {
+            result.add(new DecoratorBuildItem(KNATIVE, new AddPvcVolumeToRevisionDecorator(PvcVolumeConverter.convert(e))));
+        });
+
+        config.getAwsElasticBlockStoreVolumes().entrySet().forEach(e -> {
+            result.add(new DecoratorBuildItem(KNATIVE,
+                    new AddAwsElasticBlockStoreVolumeToRevisionDecorator(AwsElasticBlockStoreVolumeConverter.convert(e))));
+        });
+
+        config.getAzureFileVolumes().entrySet().forEach(e -> {
+            result.add(new DecoratorBuildItem(KNATIVE,
+                    new AddAzureFileVolumeToRevisionDecorator(AzureFileVolumeConverter.convert(e))));
+        });
+
+        config.getAzureDiskVolumes().entrySet().forEach(e -> {
+            result.add(new DecoratorBuildItem(KNATIVE,
+                    new AddAzureDiskVolumeToRevisionDecorator(AzureDiskVolumeConverter.convert(e))));
         });
         return result;
     }

--- a/integration-tests/kubernetes/quarkus-standard-way/src/test/java/io/quarkus/it/kubernetes/KnativeWithVolumesTest.java
+++ b/integration-tests/kubernetes/quarkus-standard-way/src/test/java/io/quarkus/it/kubernetes/KnativeWithVolumesTest.java
@@ -1,0 +1,77 @@
+package io.quarkus.it.kubernetes;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.List;
+
+import org.assertj.core.api.Condition;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.fabric8.knative.serving.v1.Service;
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.Volume;
+import io.fabric8.kubernetes.api.model.VolumeMount;
+import io.quarkus.test.ProdBuildResults;
+import io.quarkus.test.ProdModeTestResults;
+import io.quarkus.test.QuarkusProdModeTest;
+
+public class KnativeWithVolumesTest {
+
+    @RegisterExtension
+    static final QuarkusProdModeTest config = new QuarkusProdModeTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class).addClasses(GreetingResource.class))
+            .setApplicationName("knative-with-volumes-properties")
+            .setApplicationVersion("0.1-SNAPSHOT")
+            .withConfigurationResource("knative-with-volumes.properties");
+
+    @ProdBuildResults
+    private ProdModeTestResults prodModeTestResults;
+
+    @Test
+    public void assertGeneratedResources() throws IOException {
+        Path kubernetesDir = prodModeTestResults.getBuildDir().resolve("kubernetes");
+        assertThat(kubernetesDir)
+                .isDirectoryContaining(p -> p.getFileName().endsWith("knative.json"))
+                .isDirectoryContaining(p -> p.getFileName().endsWith("knative.yml"))
+                .satisfies(p -> assertThat(p.toFile().listFiles()).hasSize(2));
+
+        List<HasMetadata> kubernetesList = DeserializationUtil
+                .deserializeAsList(kubernetesDir.resolve("knative.yml"));
+
+        assertThat(kubernetesList).filteredOn(i -> "Service".equals(i.getKind())).singleElement().satisfies(i -> {
+            assertThat(i).isInstanceOfSatisfying(Service.class, s -> {
+                assertThat(s.getSpec()).satisfies(spec -> {
+
+                    assertThat(spec.getTemplate()).satisfies(template -> {
+                        assertThat(template.getSpec()).satisfies(revisionSpec -> {
+                            assertThat(revisionSpec.getVolumes()).haveAtLeastOne(new Condition<Volume>(
+                                    v -> v.getName().equals("client-crts")
+                                            && v.getSecret().getSecretName().equals("clientcerts"),
+                                    "Has secret volume named client-crts referencing secert clientcerts"));
+                            assertThat(revisionSpec.getVolumes()).haveAtLeastOne(new Condition<Volume>(
+                                    v -> v.getName().equals("client-cfg") && v.getConfigMap().getName().equals("clientconfig"),
+                                    "Has config-map named client-cfg referencing configmap clientconfig"));
+
+                            assertThat(revisionSpec.getContainers()).hasSize(1).singleElement().satisfies(c -> {
+
+                                assertThat(c.getVolumeMounts()).haveAtLeastOne(new Condition<VolumeMount>(
+                                        m -> m.getName().equals("client-crts"), "Has client-crts mount"));
+                                assertThat(c.getVolumeMounts()).haveAtLeastOne(new Condition<VolumeMount>(
+                                        m -> m.getName().equals("client-cfg"), "Has client-cfg mount"));
+
+                                assertThat(c.getPorts()).hasSize(1).singleElement().satisfies(p -> {
+                                    assertThat(p.getName()).isEqualTo("http1");
+                                });
+                            });
+                        });
+                    });
+                });
+            });
+        });
+    }
+}

--- a/integration-tests/kubernetes/quarkus-standard-way/src/test/resources/knative-with-volumes.properties
+++ b/integration-tests/kubernetes/quarkus-standard-way/src/test/resources/knative-with-volumes.properties
@@ -1,0 +1,8 @@
+# Configuration file
+quarkus.kubernetes.deployment-target=knative
+
+quarkus.knative.mounts.client-crts.path=/mnt/clientcerts
+quarkus.knative.secret-volumes.client-crts.secret-name=clientcerts
+
+quarkus.knative.mounts.client-cfg.path=/mnt/clientconfig
+quarkus.knative.config-map-volumes.client-cfg.config-map-name=clientconfig

